### PR TITLE
For connector scripts, update existing rows if changes detected

### PR DESCRIPTION
**Pull request by rudokemper**
## Goal

Refining the database writing process towards resolving https://github.com/ConservationMetrics/gc-scripts-hub/issues/22. For now, I've chosen to stick with KoboToolbox to maintain focus.

## What I changed

In looking at how we are doing DB writes, I conclude that the `safe_insert` method only inserts new rows and does nothing with existing ones. Consequently, `UniqueViolation` errors are avoided entirely, as no exceptions are raised for existing rows.

Hence, as a first attempt towards improving this, I am first checking to see if there is an existing row by running a `SELECT` query, and then if found, comparing the existing row's values with the incoming upstream data values to be inserted. If false, then insert the data as before, but I've added an `ON CONFLICT ... DO UPDATE SET` clause to handle rows that require an update.

As noted in #22, this is operating under the assumption that we will always want to update records if the upstream data source changes. So if any diff whatsoever is detected, then the row is updated.

Lastly, I find the logs in this KoboToolbox script to be rather sparse, so I added `inserted_count` and `updated_count` vars to track the number of inserts and updates, respectively, and logged those at the end.

While it seems to work, I am unsure if this is the most parsimonious or performative approach. I'm also mindful of the fact that @IamJeffG has petitioned several times that we should move to refactor these DB Writer classes as a whole, and so maybe this work is somewhat moot. Feedback welcome.

### CODE REVIEW

File: f/connectors/kobotoolbox/kobotoolbox_responses.py [hunk 1]
```diff
@@ -458,17 +459,51 @@ def _safe_insert(cursor, table_name, columns, values):
...
+
+        if existing_row and list(existing_row) == values:
+            # No changes, skip the update
+            return inserted_count, updated_count
+
         query = sql.SQL(
-            "INSERT INTO {table} ({fields}) VALUES ({placeholders})"
+            "INSERT INTO {table} ({fields}) VALUES ({placeholders}) "
+            "ON CONFLICT (_id) DO UPDATE SET {updates} "
+            "RETURNING (xmax = 0) AS inserted"
```
Comment (IamJeffG) [inline f/connectors/kobotoolbox/kobotoolbox_responses.py:485]: `xmax = 0` is postgres magic.  Requires a code comment.

File: f/connectors/kobotoolbox/kobotoolbox_responses.py [hunk 2]
```diff
@@ -458,17 +459,51 @@ def _safe_insert(cursor, table_name, columns, values):
...
+        select_query = sql.SQL("SELECT {fields} FROM {table} WHERE _id = %s").format(
+            fields=sql.SQL(", ").join(map(sql.Identifier, columns)),
+            table=sql.Identifier(table_name),
+        )
+        cursor.execute(select_query, (values[columns.index("_id")],))
+        existing_row = cursor.fetchone()
+
+        if existing_row and list(existing_row) == values:
+            # No changes, skip the update
+            return inserted_count, updated_count
```
Comment (IamJeffG) [inline f/connectors/kobotoolbox/kobotoolbox_responses.py:480]: Why do we need everything above?  Won't the INSERT below just re-write the same values and the end result is the same?

Comment (IamJeffG) [reply f/connectors/kobotoolbox/kobotoolbox_responses.py:480]: ah, is it so you can more accurately report on inserted_count, updated_count, without considering an exact overwrite to be an update?

**Review by IamJeffG — COMMENTED** [review]
I am totally on-board using `INSERT ... ON CONFLICT UDPATE`.

I will say that this is some pretty magic looking code, hard to follow.  But probably correct and as well as we can do.  An idea to consider is to insert/update all the rows in a single INSERT statement, rather than a for loop, but I don't feel strongly about this.

Major question for you: Sometimes the primary key is a hash of the values.  If the values change, then so does the primary key: there is no conflict and so could we still end up with duplicate rows?

A different side of the same question is: Why do we think we want to update old rows at all?  Do we have some reason to trust today's values more than yesterday's values of yesterday's data?

File: f/connectors/kobotoolbox/kobotoolbox_responses.py (hunk 2 above)
Comment (rudokemper) [reply f/connectors/kobotoolbox/kobotoolbox_responses.py:480]: Yep, that's correct - otherwise everything that isn't inserted is counted as updated.

We absolutely do not _need_ it for the core ETL functionality of the script, but arguably, it's a useful feature for Windmill operators to understand what their runs are doing each time. 

It could be interesting to benchmark how the script performs with and without this code for the largest Kobo dataset that we are ingesting.

Comment (IamJeffG) [reply f/connectors/kobotoolbox/kobotoolbox_responses.py:480]: I would document this, too, then.

Otherwise looks good to me!

**Comment by rudokemper** [comment]
> Major question for you: Sometimes the primary key is a hash of the values. If the values change, then so does the primary key: there is no conflict and so could we still end up with duplicate rows?

Yeah, we'll have to carefully consider our approach to hashing UUIDs for the scripts that use them. KoboToolbox (and CoMapeo) provide their own unique identifiers so we don't have this problem here. 

For the alerts script: we hash alert records based on `alert_id`, and alert metadata based on `year`, `month`, `territory_id`, `type_alert`, and `description_alerts`. 

* I can reach out to the alerts provider to confirm that `alert_id` is static and will never change. 
* For metadata, maybe we omit `type_alert` and `description_alerts`.

I realize there are still potential pitfalls here (like the provider making a mistake), but I'm not sure what else we could do to avoid those.

> A different side of the same question is: Why do we think we want to update old rows at all? Do we have some reason to trust today's values more than yesterday's values of yesterday's data?

This is the million dollar question here. 

I think there are definitely some very good reasons why we want to update old rows. I've written some of them out in #22 for each connector script. These are real scenarios that have come up in practice.

That said, I have to say that it doesn't sit well with me to entirely erase the record of what was there before. 

Imagine a scenario where a new community organization administrator gets the keys to a KoboToolbox account. They review a previously created form, and decide to modify some of the submissions for data QA reasons in accordance to their best judgment. But in so doing, they erase some things that were recorded in a specific way with good reason. I have seen this happen many times at organizations, including ones that I have worked at. (And there is of course a more alarming version of this story where a bad actor gets the keys.)

Maybe in the future we can consider version control for data records. But in the absence of that, we have a binary choice between updating and not updating, and with that binary I definitely think we should update.

**Comment by IamJeffG** [comment]
Thanks! Go ahead and merge at will.
